src/core/Akka/Actor/Futures.cs

```diff
@@ -337,28 +341,37 @@ public override int GetHashCode()
             sender = sender ?? ActorRefs.NoSender;
             var result = new TaskCompletionSource<object>();
             var a = new PromiseActorRef(provider, result, messageClassName);
-            var cancellationSource = new CancellationTokenSource();
-            cancellationSource.Token.Register(CancelAction, result);
-            cancellationSource.CancelAfter(timeout);
 
-            //var scheduler = provider.Guardian.Underlying.System.Scheduler.Advanced;
-            //var c = new Cancelable(scheduler, timeout);
-            //scheduler.ScheduleOnce(timeout, () => result.TrySetResult(new Status.Failure(new AskTimeoutException(
-            //    string.Format("Ask timed out on [{0}] after [{1} ms]. Sender[{2}] sent message of type {3}.", targetName, timeout.TotalMilliseconds, sender, messageClassName)))),
-            //    c);
+            if (timeout != TimeSpan.Zero)
+            {
+                // avoid CTS + delegate allocation if timeouts aren't needed
+                var cancellationSource = new CancellationTokenSource();
```

Comment:
Are we OK without having using semantics on TCS here?

Reply:
yeah let me take a look at the disposal code and make sure that's getting zapped